File: backend/app/services/guardrails.py
from dataclasses import dataclass


@dataclass
class GuardrailResult:
    blocked: bool
    reason: str
    response: str | None = None


class GuardrailService:
    SECURITY_BLOCK_TERMS = {
        "sql injection",
        "xss",
        "malware",
        "exploit",
        "ddos",
        "credential stuffing",
    }

    OFF_SCOPE_TERMS = {
        "write code",
        "python script",
        "movie recommendation",
        "recipe",
        "crypto trading",
    }

    DOMAIN_TERMS = {
        "gitlab",
        "handbook",
        "direction",
        "deployment",
        "marketing",
        "engineering",
        "strategy",
        "pipeline",
        "ci",
        "cd",
        "security",
        "product",
        "okr",
    }

    def check(self, question: str) -> GuardrailResult:
        lowered = question.lower()

        if any(term in lowered for term in self.SECURITY_BLOCK_TERMS):
            return GuardrailResult(
                blocked=True,
                reason="security-policy-block",
                response=(
                    "I cannot help with harmful cybersecurity misuse requests. "
                    "I can help with GitLab documentation, processes, and strategy topics."
                ),
            )

        if any(term in lowered for term in self.OFF_SCOPE_TERMS):
            return GuardrailResult(
                blocked=True,
                reason="off-scope-query",
                response=(
                    "I am scoped to GitLab Handbook and Direction knowledge. "
                    "Please ask a GitLab process, team, or strategy question."
                ),
            )

        tokens = [token for token in self._normalize(lowered) if len(token) > 2]
        if tokens:
            domain_hits = sum(1 for token in tokens if token in self.DOMAIN_TERMS)
            ratio = domain_hits / len(tokens)
            if ratio < 0.06 and len(tokens) > 4:
                return GuardrailResult(
                    blocked=True,
                    reason="low-domain-relevance",
                    response=(
                        "That looks outside this assistant's scope. "
                        "Try asking about GitLab handbook policies, direction, teams, or delivery strategy."
                    ),
                )

        return GuardrailResult(blocked=False, reason="allowed")

    def _normalize(self, text: str) -> list[str]:
        cleaned = "".join(char.lower() if char.isalnum() else " " for char in text)
        return cleaned.split()

File: nextapp/lib/guardrails.js
const SECURITY_BLOCK_TERMS = new Set([
  'sql injection',
  'xss',
  'malware',
  'exploit',
  'ddos',
  'credential stuffing',
]);

const OFF_SCOPE_TERMS = new Set([
  'write code',
  'python script',
  'movie recommendation',
  'recipe',
  'crypto trading',
]);

const DOMAIN_TERMS = new Set([
  'gitlab',
  'handbook',
  'direction',
  'deployment',
  'marketing',
  'engineering',
  'strategy',
  'pipeline',
  'ci',
  'cd',
  'security',
  'product',
  'okr',
]);

function normalize(text) {
  return text.replace(/[^a-z0-9\s]/g, ' ').split(/\s+/).filter((t) => t.length > 2);
}

/**
 * @param {string} question
 * @param {Array} history
 * @returns {{ blocked: boolean, reason: string, response?: string }}
 */
export function checkGuardrails(question, history = []) {
  const lowered = question.toLowerCase();

  if ([...SECURITY_BLOCK_TERMS].some((t) => lowered.includes(t))) {
    return {
      blocked: true,
      reason: 'security-policy-block',
      response:
        'I cannot help with harmful cybersecurity misuse requests. ' +
        'I can help with GitLab documentation, processes, and strategy topics.',
    };
  }

  if ([...OFF_SCOPE_TERMS].some((t) => lowered.includes(t))) {
    return {
      blocked: true,
      reason: 'off-scope-query',
      response:
        'I am scoped to GitLab Handbook and Direction knowledge. ' +
        'Please ask a GitLab process, team, or strategy question.',
    };
  }

  // Skip domain-ratio check for continuing conversations
  if (!history || history.length === 0) {
    const tokens = normalize(lowered);
    if (tokens.length > 4) {
      const domainHits = tokens.filter((t) => DOMAIN_TERMS.has(t)).length;
      const ratio = domainHits / tokens.length;
      if (ratio < 0.06) {
        return {
          blocked: true,
          reason: 'low-domain-relevance',
          response:
            "That looks outside this assistant's scope. " +
            'Try asking about GitLab handbook policies, direction, teams, or delivery strategy.',
        };
      }
    }
  }

  return { blocked: false, reason: 'allowed' };
}
